x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -410,25 +473,43 @@ private void onSearchResponse(SearchResponse searchResponse) {
     private void onBulkResponse(BulkResponse response, JobPosition position) {
         stats.markEndIndexing();
         try {
-            ActionListener<SearchResponse> listener = ActionListener.wrap(this::onSearchResponse, this::finishWithSearchFailure);
             // TODO probably something more intelligent than every-50 is needed
             if (stats.getNumPages() > 0 && stats.getNumPages() % 50 == 0) {
                 doSaveState(IndexerState.INDEXING, position, () -> {
-                    nextSearch(listener);
+                    nextSearch();
                 });
             } else {
-                nextSearch(listener);
+                nextSearch();
             }
         } catch (Exception e) {
             finishWithIndexingFailure(e);
         }
     }
 
-    private void nextSearch(ActionListener<SearchResponse> listener) {
+    protected void nextSearch() {
+        if (maximumRequestsPerSecond > 0 && lastDocCount > 0) {
+            TimeValue executionDelay = calculateThrottlingDelay(maximumRequestsPerSecond, lastDocCount, lastSearchStartTimeNanos,
+                    getTimeNanos());
+
+            if (executionDelay.duration() > 0) {
+                logger.debug("throttling job [{}], wait for {} ({} {})", getJobId(), executionDelay, maximumRequestsPerSecond,
+                        lastDocCount);
+                scheduledNextSearch = threadPool.schedule(() -> triggerNextSearch(), executionDelay, executorName);
```

Comment:
Since this is not `synchronized`, I am not sure if `private synchronized void reQueueThrottledSearch()` or `private synchronized void triggerThrottledSearchNow()` actually work :/. The `scheduledNextSearch` could get set out from under those blocks. Unless we are ONLY worried about `reQueueThrottledSearch` and `triggerThrottledSearchNow` conflicting.

Reply:
good point. Synchronized for re-throttle and stop is required, because there could be parallel calls to it. `nextSearch` only runs in the indexer thread, so its never called more than once at the same time. 

The purpose of `triggerThrottledSearchNow` is to avoid a long waiting time between calling stop and the actual stop. For this case the idea is to trigger the search immediately. However its still possible that `stop` is called but the search ends in the wait queue.

I will revisit it. Maybe its better to cancel the search request instead of re-queuing, e.g. if `stop` is called, because of cluster trouble.